### PR TITLE
[Filebeat] Update Elasticsearch Filebeat module to include cluster.uuid in audit logs

### PR DESCRIPTION
**Pull request by sayden**
WIP. Change pipeline to include UUID in audit logs.

### CODE REVIEW

**Comment by andresrc** [comment]
Missing changelog entry

File: filebeat/module/elasticsearch/audit/test/test-audit.log [hunk 1]
```diff
@@ -5,3 +5,5 @@
...
+{"timestamp": "2021-02-10T14:24:38,466Z", "type": "server", "level": "INFO", "component": "o.e.x.i.a.TransportPutLifecycleAction", "cluster.name": "docker-cluster", "node.name": "d739441e55ef", "message": "adding index lifecycle policy [kibana-event-log-policy]", "cluster.uuid": "qb5P5M_QSbWRsE1xA9MDKA", "node.id": "lz9sNb0TRCy9PO99_5aOkA"  }
+{"type":"server","timestamp":"2021-02-10T14:34:34,598Z","level":"INFO","component":"o.e.x.i.IndexLifecycleTransition","cluster.name":"docker-cluster","node.name":"d739441e55ef","message":"moving index [ilm-history-3-000001] from [{\"phase\":\"hot\",\"action\":\"unfollow\",\"name\":\"wait-for-yellow-step\"}] to [{\"phase\":\"hot\",\"action\":\"rollover\",\"name\":\"check-rollover-ready\"}] in policy [ilm-history-ilm-policy]","cluster.uuid":"qb5P5M_QSbWRsE1xA9MDKA","node.id":"lz9sNb0TRCy9PO99_5aOkA"}
```
Comment (ycombinator) [inline filebeat/module/elasticsearch/audit/test/test-audit.log:9]: @sayden Are you sure these are audit log lines? They look like ES server log lines — notice the `"type": "server"` in them. How did you generate them?

**Review by ycombinator — COMMENTED** [review]
Left a question about the new test fixture entries.

**Comment by sayden** [comment]
I was actually checking if everything is ok in the CI. I was expecting a red flag in CI, because I have 2 new input lines and no generated output but... everything is green...

**Comment by ycombinator** [comment]
> I was actually checking if everything is ok in the CI. I was expecting a red flag in CI, because I have 2 new input lines and no generated output but... everything is green...

Well, we should add log lines from the ES audit log to the audit log test file and then check if CI is green. 😄 That's the sure shot way of knowing that the `elasticsearch/audit` fileset can actually process Elasticsearch audit logs.  

If it's going green on server log lines, we might have a different issue!

How did you generate these lines? Did you enable the ES audit log and then grab them from there? If not, you'll want to do that so the lines in the test file are actual ES audit logs. Note that the ES audit log is disabled by default and it's not supported by the Basic license.

**Comment by sayden** [comment]
I found the issue. Long story is that I got some audit lines but no "_expected_" events were being generated. So I duplicated a server input log line to see if a _expected_ event was generated and it was. So I bring the line working on the `server` metricset to audit and nothing was generated again. My last attempt tonight was to upload this to CI to see if it complain... but it didn't :sweat: 

I have just discovered that the problem is that a `processor` in Filebeat (not ingest pipeline) can drop a full event silently. Later, Beats tests don't check that the number of input log lines != number of output events (probably, because of multiline logs). So the generation and the test was failing silently.

**Comment by ycombinator** [comment]
> Later, Beats tests don't check that the number of input log lines != number of output events (probably, because of multiline logs). So the generation and the test was failing silently.

Oof, if this is what's going on then I wish there was a way to configure Filebeat tests that they are for multiline logs so this check is only skipped in those cases. Anyway, out of scope for this PR. Good find on why CI gave us a green result here.

**Comment by ycombinator** [comment]
> Later, Beats tests don't check that the number of input log lines != number of output events 

Wait, are you sure about this? Maybe I'm looking in the wrong place but I found this line in the test runner code: https://github.com/elastic/beats/blob/7b6f338852482838189567611dd68dd1fcc078ca/filebeat/tests/system/test_modules.py#L203

**Comment by ycombinator** [comment]
Okay, I know what's going on. The test runner is working as expected. It compares the number of indexed objects in Elasticsearch with the number of entries in the expected JSON file. In this case, the `drop` processor in the `elasticsearch/audit` fileset's ingest pipeline is (correctly) rejecting the `"type": "server"` lines so those two new events never get indexed into Elasticsearch. Since there are no new lines in the expected JSON file either, the length of objects in ES matches up with the length of lines in the expected JSON file.

Basically, we need to have actual audit log lines in the test file so they have `"type": "audit"` (not `"type": "server"). That would have skipped the `drop` processor and the lines would've been indexed into Elasticsearch. Then, because there still weren't any corresponding new lines in the expected JSON file, the length check would've failed the test.

**Comment by ycombinator** [comment]
Okay, I'm going back to my PTO now. Please add actual ES audit log lines to the test file and let's see where that gets us. I'll be back on Tuesday and will review this PR then.

**Comment by sayden** [comment]
Oh! Good catch. But now I wonder if we are not skipping some log lines silently in other modules (with processors / pipelines not working as expected)

Anyways, I actually should have look carefully to the audit log event lines before attempting anything because they don't seem to include a `cluster.uuid`:

```json
{
    "type": "audit",
    "timestamp": "2021-02-11T15:36:53,065+0100",
    "node.id": "ANckieszR-WKu2banH-2dg",
    "event.type": "rest",
    "event.action": "anonymous_access_denied",
    "origin.type": "rest",
    "origin.address": "[::1]:41168",
    "url.path": "/_security/api_key",
    "request.method": "POST",
    "request.id": "e7JqDUiQQRqzPuKag1QTLg"
}
```

**Comment by ycombinator** [comment]
> But now I wonder if we are not skipping some log lines silently in other modules (with processors / pipelines not working as expected)

As long as we add expected lines to go 1-1 with test lines, the tests will do the right thing. Part of the problem here was that no expected lines were added. I really don't think the test runner is the problem here.

Also, the `elasticsearch/audit` pipelines _is_ working as expected. It is intended to only parse `"type": "audit"` logs so it rejects any other types, e.g. `"type": "server"`.

**Comment by ycombinator** [comment]
> Anyways, I actually should have look carefully to the audit log event lines before attempting anything because they don't seem to include a `cluster.uuid`.

Yup, another benefit of including the actual lines from the log we're testing 😉. Good find — maybe check with the ES team next about this? It's an enhancement (as opposed to a breaking change), hopefully an easy one, and we have precedent in other ES JSON logs for including a `cluster.uuid` in them. So they should be able to include it in audit JSON logs as well.

**Comment by ycombinator** [comment]
Okay, now I'm _really_ going back to my PTO 😄 👋.

**Comment by sayden** [comment]
Closing this as the field isn't present finally.
